x/observer/keeper/msg_server_add_block_header.go

```diff
@@ -61,8 +43,20 @@ func (k msgServer) AddBlockHeader(goCtx context.Context, msg *types.MsgAddBlockH
 		return nil, cosmoserrors.Wrap(types.ErrBlockAlreadyExist, hashString)
 	}
 
+	bhs, found := k.Keeper.GetBlockHeaderState(ctx, msg.ChainId)
+	if found && bhs.EarliestHeight > 0 && bhs.EarliestHeight < msg.Height {
+		phash, err := msg.Header.ParentHash()
+		if err != nil {
+			return nil, cosmoserrors.Wrap(types.ErrNoParentHash, err.Error())
+		}
+		_, found = k.GetBlockHeader(ctx, phash)
```

Comment:
Just in case, maybe we can make it more strict
```
header, found := k.GetBlockHeader(ctx, phash)
if !found {
	return nil, cosmoserrors.Wrap(types.ErrNoParentHash, "parent block header not found")
}
if header.Height+1 != msg.Height {
	return nil, cosmoserrors.Wrapf(types.ErrParentHeightMismatch, "invalid height, expected: %d, got: %d", header.Height+1, msg.Height)
}
if bytes.Compare(header.Hash, phash) != 0 {
	return nil, cosmoserrors.Wrapf(types.ErrParentHashMismatch, "parent hash mismatch, expected: %s, got: %s", msg.BlockHash, phash)
}
```

Reply:
revised a bit